x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/MatchPhrase.java

```diff
@@ -0,0 +1,379 @@
+        super(
+            source,
+            matchPhraseQuery,
+            options == null ? List.of(field, matchPhraseQuery) : List.of(field, matchPhraseQuery, options),
+            queryBuilder
+        );
+        this.field = field;
+        this.options = options;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    private static MatchPhrase readFrom(StreamInput in) throws IOException {
+        Source source = Source.readFrom((PlanStreamInput) in);
+        Expression field = in.readNamedWriteable(Expression.class);
+        Expression query = in.readNamedWriteable(Expression.class);
+        QueryBuilder queryBuilder = null;
+        queryBuilder = in.readOptionalNamedWriteable(QueryBuilder.class);
+        return new MatchPhrase(source, field, query, null, queryBuilder);
+    }
+
+    @Override
+    public final void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(field());
+        out.writeNamedWriteable(query());
+        out.writeOptionalNamedWriteable(queryBuilder());
+    }
+
+    @Override
+    protected TypeResolution resolveParams() {
+        return resolveField().and(resolveQuery()).and(resolveOptions()).and(checkParamCompatibility());
+    }
+
+    private TypeResolution resolveField() {
+        return isNotNull(field, sourceText(), FIRST).and(
+            isType(field, FIELD_DATA_TYPES::contains, sourceText(), FIRST, "keyword, text, boolean, date, date_nanos, ip, version")
+        );
+    }
+
+    private TypeResolution resolveQuery() {
+        return isType(query(), QUERY_DATA_TYPES::contains, sourceText(), SECOND, "keyword, text").and(
+            isNotNullAndFoldable(query(), sourceText(), SECOND)
+        );
+    }
+
+    private TypeResolution checkParamCompatibility() {
+        DataType fieldType = field().dataType();
+        DataType queryType = query().dataType();
+
+        // Field and query types should match. If the query is a string, then it can match any field type.
+        if ((fieldType == queryType) || (queryType == KEYWORD) || (queryType == TEXT)) {
+            return TypeResolution.TYPE_RESOLVED;
+        }
+
+        return new TypeResolution(formatIncompatibleTypesMessage(fieldType, queryType, sourceText()));
+    }
+
+    private TypeResolution resolveOptions() {
+        if (options() != null) {
+            TypeResolution resolution = isNotNull(options(), sourceText(), THIRD);
+            if (resolution.unresolved()) {
+                return resolution;
+            }
+            // MapExpression does not have a DataType associated with it
+            resolution = isMapExpression(options(), sourceText(), THIRD);
+            if (resolution.unresolved()) {
+                return resolution;
+            }
+
+            try {
+                matchPhraseQueryOptions();
```

Comment:
the `resolveOptions` function could use a refactor - because most of it is duplicated in `Match`, `MultiMatch`, `QueryString` and here. But I am okay with doing that as a follow up.